From 0da627d4a92eb13f93cd8af89cf6daf33d4671c5 Mon Sep 17 00:00:00 2001
From: Cristian Ciocaltea <cristian.ciocaltea@collabora.com>
Date: Thu, 19 Jun 2025 23:57:01 +0300
Subject: [PATCH 107/108] [WIP-FRL] drm/rockchip: vop2: Add HDMI 2.1 FRL
 support

TODO: this has been "borrowed" from downstream code, although it might
not really be required.

Signed-off-by: Cristian Ciocaltea <cristian.ciocaltea@collabora.com>
---
 drivers/gpu/drm/rockchip/rockchip_vop2_reg.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c b/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c
index 5a3d8c834ca9..74b6275da45d 100644
--- a/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c
+++ b/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c
@@ -1578,6 +1578,15 @@ static unsigned long rk3588_calc_cru_cfg(struct vop2_video_port *vp, int id,
 		 */
 		*if_pixclk_div = 2;
 		*if_dclk_div = 4;
+
+		if (dclk_rate > 600000000) {
+			dclk_rate = dclk_rate >> 1;
+			*if_pixclk_div = 1;
+			*if_dclk_div = 2;
+		}
+
+		*dclk_out_div = 1;
+
 	} else if (vop2_output_if_is_edp(id)) {
 		/*
 		 * edp_pixclk = edp_dclk > dclk_core
-- 
2.34.1

